[linters] Install tools based on versions in package.json

chalin opened the pull request:
I'm having issues running certain spec checks from the website because it's working from a spec submodule. This PR:

- Switches to using `npx` to run npm commands rather than hardcoded paths
- Uses `npm install` to install all necessary NPM-based tools
  - Adds a `package.json` file for this purpose -- with `markdownlint-cli` pinned to `v0.31.0` as was done in #2320

IMHO, this is a clean and conventional way to deal with NPM packages. WDYT?

/cc @austinlparker

chalin commented on CONTRIBUTING.md:
````diff
@@ -92,11 +93,10 @@ wide. There are tools that can do it for you effectively. Please submit proposal
+To check for style violations, run:
 
 ```bash
-make install-markdownlint # install markdownlint if not installed
````
Btw, the `install-markdownlint` target doesn't exist. It must have been removed at one point.

chalin commented on Makefile:
```diff
@@ -24,15 +21,10 @@ misspell:	$(MISSPELL)
-$(MARKDOWN_TOC):
-	npm install markdown-toc
+markdown-link-check:
+	@if ! npm ls markdown-link-check; then npm install; fi
```
If we always run `npm install` first before any check-tool targets, then we don't need this check. I added it to maintain backwards compatibility with the behavior of the make target as it was before this PR.

If you prefer, I can remove the `@if ...` line here and for the other two targets.

joaopgrassi replied:
What about failing/not running if it's not installed? That's much simpler, I guess? Now that there's `npm install` in the contributing guide, I'm not sure if it's necessary to install here. It does change the existing behavior so..

chalin replied:
Yes, that would be simpler, and it would be the solution I prefer. I've pulled out two of the three `@if ! npm ls foo; then npm install; fi` lines, because two of them were actually unnecessary (I wasn't aware that `npx` would install a package if it wasn't already installed). I've kept the line for `markdownlint` because `npx` fails otherwise (because it doesn't know about the CLI package).

chalin commented on package.json:
```diff
@@ -0,0 +1,7 @@
+  "devDependencies": {
+    "markdown-link-check": "^3.9.3",
+    "markdown-toc": "^1.2.0",
+    "markdownlint-cli": "0.31.0"
```
As I mentioned, the version of `markdownlint-cli` is pinned here, as was done in #2320.

joaopgrassi replied:
Maybe we can pin all of them? Better be safe than random build failures? 🤣

yurishkuro reviewed (approved):
❤️

anuraaga commented on Makefile:
```diff
@@ -24,15 +21,10 @@ misspell:	$(MISSPELL)
-	npm install markdown-toc
+markdown-link-check:
+	@if ! npm ls markdown-link-check; then npm install; fi
+	@for f in $(ALL_DOCS); do npx markdown-link-check --quiet --config .markdown_link_check_config.json $$f; done
```
I think `npx` will install the latest version of the package if someone forgot to run `npm install`, ignoring `package.json`, while `npm run` would avoid this to guarantee the version used is what's in the `package.json`. Though alternatively, what do you think about removing `package.json` / `npm install` steps and adding the version in the `npx` command, e.g. `npx markdown-link-check@3.9.3`?

joaopgrassi replied:
Another alternative to npx is adding a script in `package.json` to run the tool locally installed. 

```
"scripts": {
    "markdown-lint": "./node_modules/.bin/markdown-link-check"
  }
  ```
Then this line would contain `npm run markdown-lint` instead of `npx..`. 

Now that we have a `package.json`, it "unlocks" this alternative.

chalin replied:
> I think `npx` will install the latest version of the package if someone forgot to run `npm install`, ignoring `package.json`,

According to my tests, unless you specify an explicit package version on the command line, `npx` will install the package version specified in `packages.json`.

> what do you think about removing `package.json` / `npm install` steps and adding the version in the `npx` command, e.g. `npx markdown-link-check@3.9.3`?

IMHO, the proper place to specify a package version is in the `packages.json` file rather than the command line. Also, `npx markdownlint` doesn't work because we need it's CLI package too:

```console
$ npx markdownlint
npm ERR! could not determine executable to run
```

So I think that the cleanest and most standard solution is to have a `packages.json` file to specify the packages used by this repo.

chalin replied:
@anuraaga - you're right, `npx` will fetch the latest. I prefer @joaopgrassi's suggestion of just requiring users to run `npm install`.

joaopgrassi commented on package.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "devDependencies": {
```
I think we should also consider saving the [lock file](https://docs.npmjs.com/cli/v8/configuring-npm/package-lock-json), to make sure everyone gets the exact same dependencies?

chalin replied:
For the use that we're making of these packages, I'd vote against saving the lock file. It is too much of a pain to deal with merge conflicts.

joaopgrassi replied:
Sure. But then, it would make sense to pin the versions of all packages, no (like done for `markdownlint-cli`)? Otherwise new versions can be downloaded and we can't know the effect, specially in CI.

chalin replied:
Actually, I unpinned `markdownlint-cli` because `0.31.0` _is_ the latest version. I'd assume that the packages that we're using are following semver and so if there's a breaking change it'll be introduced in the next major version (or minor version in the case of `markdownlint-cli`), for which we'll need to update the package.json file.

Again, we're not building a binary, we're just checking markdown files. I don't think that we need to pin package versions (unless something is known to be broken). We can rely on semver working for us. My 2 cents.

joaopgrassi replied:
Sure, fair enough. I just wanted to point out this, since I had my share of issues with npm packages not respecting that.

joaopgrassi reviewed (commented):
This is great! I just left a few small things.

chalin commented on Makefile:
```diff
@@ -24,15 +21,12 @@ misspell:	$(MISSPELL)
+markdown-link-check:
+	@for f in $(ALL_DOCS); do \
+		npx --no -- markdown-link-check --quiet --config .markdown_link_check_config.json $$f \
+			|| exit 1; \
```
Note that link check will now fail if the package isn't installed or on the first broken link that it finds.

joaopgrassi reviewed (approved):
Looks good to me! I only have one concern about pinning package versions to avoid surprises (as we had a few days ago). But I'll leave that to others to consider as well.

chalin commented on package.json:
```diff
@@ -0,0 +1,7 @@
+  "devDependencies": {
+    "markdown-link-check": "^3.9.3",
+    "markdown-toc": "^1.2.0",
+    "markdownlint-cli": "^0.31.0"
```
> ... change is backwards compatible to the library itself, it's not for consumers ...

Oh, that totally makes sense. Ok, I've pinned it for `markdownlint-cli`.

chalin commented on .github/workflows/checks.yaml:
```diff
@@ -14,7 +14,9 @@ jobs:
     - name: run markdownlint
-      run: make markdownlint
+      run: |
+        npm install
```
> Does it make sense to add a separate step to install the dependencies?

I'm not 100% sure that I understand (the scope/context of) your question, and I'm not an expert on GH workflows, but this is how I've seen other projects call `npm install`.

chalin replied:
> Also, not sure it makes sense for this small thing, but we could consider caching the packages https://docs.github.com/en/actions/advanced-guides/caching-dependencies-to-speed-up-workflows

Thanks for the link. That would make sense. Maybe this optimization can be for a followup PR?

chalin replied:
> Does it make sense to add a separate step to install the dependencies?

Done

joaopgrassi replied:
Yes, sounds good 👍

chalin commented:
@joaopgrassi - thanks for the feedback! I've made the changes that you suggested. PTAL

joaopgrassi commented on .github/workflows/checks.yaml:
```diff
@@ -13,6 +13,9 @@ jobs:
     - name: check out code
       uses: actions/checkout@v2
 
+    - name: npm install
```
I would call it something like "Install check dependencies" or just "Install dependencies"

joaopgrassi replied:
This is the text that is shown in the Actions page
![image](https://user-images.githubusercontent.com/5938087/153616059-4e2f73e7-cd1f-410b-80c0-b4b7ca4add73.png)

chalin replied:
Done

joaopgrassi reviewed (approved):
LGTM

joaopgrassi commented:
Hi @chalin, I think this is ready right? @anuraaga maybe you can take another look? Would be a pity if this is closed due to inactivity

Oberon00 commented on CONTRIBUTING.md:
````diff
@@ -92,11 +93,10 @@ wide. There are tools that can do it for you effectively. Please submit proposal
 ```bash
-make install-markdownlint # install markdownlint if not installed
-make markdownlint # run at the root of the repo
+make markdownlint
````
No npm install required here?

chalin replied:
I just submitted a commit that adjust the wording from earlier, making it clear that `npm install` needs to be run first before all checks. PTAL

Oberon00 commented on CONTRIBUTING.md:
````diff
@@ -45,7 +45,8 @@ The Specification has a number of tools it uses to check things like style,
 ```bash
-make check
+npm install # get the tools
+make check  # run all checks
````
Can't `make check` still do that automatically, i.e. can we make `make` call `npm install` if required (e.g. by checking for presence of node_modules subfolder or calling npm install with options that cause it to not upgrade already installed stuff)?

chalin replied:
Yes it _can_, but part of the point of this PR has been to separate concerns by having NPM package installation be consolidated (with packages listed in `packages.json`), with installation purposefully being a separate step from checks.

(Also, IMHO, eventually many of the Make targets could be replaced by NPM scripts. But that's a topic for another day.)

joaopgrassi replied:
I think with the latest change in `CONTRIBUTING.md` is clear enough that one should first install the packages in order to run the checks. This also removes some complexity in the make file.

Oberon00 replied:
As a user of the spec build (which includes most editors of the spec), this separation of concerns does not add any value for me, does it?
And isn't it an implementation detail whether some tool is installed via npm or some other means? The spec is not a Node.js project, so it's unreasonable to assume all supporting tools will be installed via npm.

Is it possible that this PR makes work easier for maintainers of the spec build, but makes work (ever so slightly) harder for everybody else working on the spec?

yurishkuro replied:
@Oberon00 what would you suggest? If a casual contributor does not have a Node.js installation, then we wouldn't be able to fix that via the Makefile.

As a side note, it would be cool to enable GH Workspaces (or whatever it's called), so that casual contributors could propose PRs directly from a web interface using a VM with prebuilt tooling env.

chalin replied:
@Oberon00 - I'm not against preserving the old behavior, which is to allow a dev to invoke any make target without having to run `npm install` beforehand. In fact, that's how I had it setup originally, e.g.:

```make
markdown-link-check:
	@if ! npm ls markdown-link-check; then npm install; fi
	...
```

For details, see this https://github.com/open-telemetry/opentelemetry-specification/pull/2323#discussion_r801084679.

> As a user of the spec build (which includes most editors of the spec), this separation of concerns does not add any value for me, does it? ... Is it possible that this PR makes work easier for maintainers of the spec build, but makes work (ever so slightly) harder for everybody else working on the spec?

IMHO, the most important contribution in this PR is the introduction of `packages.json` to replace lines like these: https://github.com/open-telemetry/opentelemetry-specification/blob/21a3191bced91e5de2396784951b9e343a1d02c0/Makefile#L60-L61

It's better to list packages and their versions (pinned or not) in a `packages.json` file so that standard tools can process them and, for example, report vulnerabilities and security issues. As you might know, there exists bots to do that automatically, scanning repos and opening issues when packages need updating -- such as [dependabot](https://github.com/dependabot).

To me its 50/50 whether a make target installs dependencies if they aren't already installed, as long as the command used is `npm install` -- i.e., that we install packages as specified in the `packages.json` file.

Up to y'all to decide. Let me know.

chalin replied:
> As a side note, it would be cool to enable GH Workspaces

@yurishkuro et all: have you tried GitPod.io? You can open this repo using the following link, https://gitpod.io/#https://github.com/open-telemetry/opentelemetry-specification, without any further configuration and it will immediately run `make`.

I just pushed a commit (https://github.com/open-telemetry/opentelemetry-specification/pull/2323/commits/21d4389c78890b0ac0183f70cfb8dc5ad0837544) to ensure that if/when this PR is merged, then we can continue using GitPod.io without custom settings (not that it would be much work).

yurishkuro commented on Makefile:
```diff
@@ -28,15 +25,12 @@ misspell:	$(MISSPELL)
-$(MARKDOWN_TOC):
-	npm install markdown-toc
+markdown-link-check:
+	@for f in $(ALL_DOCS); do \
```
Not a ding on this PR, but I found it extremely annoying that the checks are done in this manner, one file at a time (paying the cost of invoking the whole npx chain every time). I think most of the tools would accept a list of files, which would make the linting faster and less verbose.

chalin replied:
(Agreed. FYI, I'm looking into an alternative tool. But that's possibly for another PR.)

chalin commented:
Hi all: I've restored the old target behavior by using install-checks. I see this as win-win, WDYT @Oberon00, @joaopgrassi,  @yurishkuro @carlosalberto and others? While conceptually less clean, this addresses the concerns that @Oberon00 raised. My hope is that we can now have enough of a consensus to get this PR merged.

joaopgrassi reviewed (approved):
Both approaches are fine to me. I'm no expert in Make, but the build passes and I don't see anything clearly wrong. 👍

chalin commented on CONTRIBUTING.md:
````diff
@@ -42,7 +42,13 @@ one that is needlessly restrictive and complex.
+spelling and link validity. Before using the tools, run:
+
+```bash
+npm install
````
I'd leave this as the recommended workflow.

chalin replied:
@joaopgrassi - if you agree, could yo mark this as resolved?

joaopgrassi replied:
Yes, sounds good. I can't resolve it though, so feel free to :)

chalin commented:
Thanks for the feedback @joaopgrassi.
Is this ready to merge, before the bot closes it!?! 😉 😄

yurishkuro commented:
I am resolving discussions and merging. Feel free to open new tickets if changes are needed, but in the current state the PR is an overall improvement.

tigrannajaryan commented:
@chalin  After this change I can no longer do `make markdownlint` on my Mac. I get:

```
$ make markdownlint
/Users/tnajaryan/work/repos/opentelemetry-specification
└─┬ markdownlint-cli@0.31.0
  └── markdownlint@0.25.1 

./CHANGELOG.md
npx: installed 203 in 6.874s
npx: command not found: markdownlint-cli
make: *** [markdownlint] Error 1
```

Any advice on what I need to do to make this work locally?

yurishkuro commented:
Indeed - why did CI pass then?

It works for me after this patch:
```diff
diff --git a/Makefile b/Makefile
index e0d64e4..755e8e6 100644
--- a/Makefile
+++ b/Makefile
@@ -58,7 +58,7 @@ markdownlint:
        @if ! npm ls markdownlint; then npm install; fi
        @for f in $(ALL_DOCS); do \
                echo $$f; \
-               npx --no -p markdownlint-cli markdownlint -c .markdownlint.yaml $$f \
+               npx --no -p markdownlint markdownlint -c .markdownlint.yaml $$f \
                        || exit 1; \
        done
```

however it seems to be re-doing npm install for each file:
```
./specification/compatibility/opencensus.md
npx: installed 203 in 4.611s
./specification/compatibility/openmetrics.md
npx: installed 203 in 4.363s
./specification/compatibility/opentracing.md
npx: installed 203 in 4.437s
./specification/context/api-propagators.md
npx: installed 203 in 4.312s
./specification/context/context.md
```

yurishkuro commented:
I suggest we revert this PR until it can be properly fixed

chalin commented:
Checking. Yes, adding an `.nvmrc` is a good idea. I can do that if you'd like.

chalin commented:
Indeed, possibly a Node version issue.

> @chalin After this change I can no longer do `make markdownlint` on my Mac. I get:
> 
> ```
> $ make markdownlint
> /Users/tnajaryan/work/repos/opentelemetry-specification
> └─┬ markdownlint-cli@0.31.0
>   └── markdownlint@0.25.1 
> 
> ./CHANGELOG.md
> npx: installed 203 in 6.874s
> npx: command not found: markdownlint-cli
> make: *** [markdownlint] Error 1
> ```
> 
> Any advice on what I need to do to make this work locally?

It works fine for me locally. I'm using the latest LTS version of Node:

```console
$ nvm current
v16.13.0
```

You?

yurishkuro commented:
Confirmed, works w/ v16, but not with v14.

tigrannajaryan commented:
v16 works for me too.
